### ISSUE TITLE
ol.Map.forEachLayerAtPixel() not executing callback consistently when renderer = 'canvas'

### ISSUE THREAD

**Issue by Patchez** [openlayers/openlayers#6402]
`ol.Map.forEachLayerAtPixel()` is not executing callback consistently when `renderer = 'canvas'` even when the layerFilter returns true.  Changing the `renderer` to `'webgl'` resolves this issue.

Looking through the source code, we found some odd differences between the code for the canvas renderer and the webgl renderer when it came to `forEachLayerAtPixel`.

**Comment by marcjansen** [comment]
Which differences?

**Comment by Patchez** [comment]
We're currently working on a pull request with changes that we think resolve this.  Should have it submitted shortly.

**Comment by glenselle** [comment]
Just to add some more info on this one. When clicking on the map we perform an `ol.Map.forEachLayerAtPixel()` call with the pixel location at the click event. In the filter callback, we return a boolean based on a condition we're checking. Execution of this callback is intermittent.

To reproduce the issue, simply return `true` in the `ol.Map.forEachLayerAtPixel()`filter callback every time. When the canvas renderer is used, the callback to handle a match will be intermittently called. As soon as the renderer is changed to `webgl`, the callback is called every time as would be expected since the filter is hard-coded to return `true` every time. 

Below is the code that is intermittent:

```
let raster = new ol.layer.Tile({
  title: 'OSM Street',
  source: new ol.source.OSM({
    layer: 'osm'
  })
})
let map = window.olMap = new ol.Map({
  layers: [raster],
  target: 'map',
  controls: [],
  renderer: 'canvas'
})
map.on('click', function (e) {
  console.log('CLICK')

  map.forEachLayerAtPixel(e.pixel, function (layer) {
    console.warn('CALLBACK')
  }, this, function (layer) {
    console.log('FILTER')
    return true
  })
}
```

<img width="472" alt="screen shot 2017-01-26 at 12 20 00 pm" src="https://cloud.githubusercontent.com/assets/1056932/22344602/54957fb2-e3c2-11e6-9690-e8c86419af1d.png">

**Comment by glenselle** [comment]
It appears a JSfiddle doing this isn't able to repro the issue. We'll continue to delve further into what we might be doing with our map that creates the issue.

http://jsfiddle.net/8o5954k7/

**Comment by Patchez** [comment]
So after looking into this further we found that the jsFiddle above can reproduce the issue we are seeing by using 'https://openlayers.org/en/v3.20.1/build/ol-debug.js' as the source of openlayers.  However we see this issue on our end even if we are not using the ol-debug version.  In this JSfiddle, clicking on the bottom right of the map will not result in the callback being called while clicking in the top left will call the callback.  This is identical behavior to the issue we are seeing.

http://jsfiddle.net/7q3k0vv5/

**Comment by glenselle** [comment]
@marcjansen @ahocevar Are you also able to see the behavior with the above JSfiddle? I'm fairly confident our PR doesn't address the issue correctly, but we're definitely seeing a bug in OL3.

**Comment by Patchez** [comment]
When you click in an area that does not trigger the callback, [`ol.transform.apply(this.coordinateToCanvasPixelTransform, coordinate.slice()) `](https://github.com/openlayers/openlayers/blob/d6a57968fd2b549dd598b76b7993aeae45004410/src/ol/renderer/canvas/intermediatecanvas.js#L132 ) is returning values that are way outside of the valid pixel range (e.g. `frameState.size === [1440, 760]` but `pixel === [2837.4999999925494, 1387.999999998603]`).  This results in [`imageData`](https://github.com/openlayers/openlayers/blob/d6a57968fd2b549dd598b76b7993aeae45004410/src/ol/renderer/canvas/intermediatecanvas.js#L142) becoming [0,0,0,0] and the following conditional to return `undefined`.

**Comment by ahocevar** [comment]
Duplicate of #6250, was fixed with #6251. Fix will be in the next release.
